novawallet/Modules/Vote/Governance/Model/ReferendumLocal.swift

```diff
@@ -99,14 +99,69 @@ struct SupportAndVotesLocal {
 
     func isPassing(at block: BlockNumber) -> Bool {
         guard
-            let approvalThreshold = approvalFunction?.calculateThreshold(for: block),
-            let supportThreshold = supportFunction?.calculateThreshold(for: block),
-            let approvalFraction = approvalFraction else {
+            let approvalFunction,
+            let supportFunction,
+            let approvalFraction,
+            let approvalThreshold = approvalFunction.calculateThreshold(for: block),
+            let supportThreshold = supportFunction.calculateThreshold(for: block)
+        else {
             return false
         }
 
         return approvalFraction >= approvalThreshold && supportFraction >= supportThreshold
     }
+
+    func projectPassing(
+        currentBlock: BlockNumber,
+        since: BlockNumber,
+        period: Moment,
+        confirmPeriod: Moment?
+    ) -> VoteProjectionResult {
+        guard
+            currentBlock < (since + period),
+            let approvalFraction,
+            let approvalDelay = approvalFunction?.delay(for: approvalFraction),
+            let supportDelay = supportFunction?.delay(for: supportFraction),
+            let approvalDelayedThreshold = approvalFunction?.calculateThreshold(for: approvalDelay),
+            let supportDelayedThreshold = supportFunction?.calculateThreshold(for: supportDelay),
+            approvalFraction >= approvalDelayedThreshold,
+            supportFraction >= supportDelayedThreshold
+        else {
+            return .notPassing
+        }
+
+        let confirmingDelay = max(approvalDelay, supportDelay)
+
+        let confirmingBlockDecimal = (
+            Decimal(since)
+                + confirmingDelay
+                * Decimal(period)
+        ).ceil() as NSDecimalNumber
+
+        guard confirmingBlockDecimal.intValue > 0 else {
+            return .notPassing
+        }
+
+        let confirmingBlock = BlockNumber(confirmingBlockDecimal.intValue)
+
+        if passing(
```

Comment:
Don't see why we really need another ```passing``` function. It look umbigious (whether it current passing or projected passing). I would just write inline ```confirmingBlock < since + period```